src/librustc_mir/transform/generator.rs

```diff
@@ -369,6 +402,12 @@ fn locals_live_across_suspend_points<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
             // Mark locals without storage statements as always having live storage
             live_locals.union(&ignored.0);
 
+            if !movable {
+                // For immovable generators we consider borrowed locals to always be live.
```

Comment:
You could also only promote locals that are borrowed across a yield.

Reply:
I'd need a liveness analysis which properly handles borrows then. I'm waiting on NLL before I do that. My plan is to have a pass which optimizes `Storage*` instructions based on NLL and then let the generator transformation only use the dataflow analysis on those instructions.